README.md
# Minimalistic 4x6 (real 3x5) pixels font for dot matrix displays
Full ASCII and some KOI8-R symbols supported.
Main source file: "font_4x6_src.py".



ChangeLog
2018.01.09:
 + start project

2018.01.10:
 + add some symbols (#, $, &, ´)
 * convert 4x6 to 6*4 bytearays

2018.01.31:
 * update big 'O'



font_4x6.py
#!/usr/bin/env python3
# -*- coding: UTF8 -*-

font_4x6 = {
  0x0020: b"\x00\x00\x00\x00", # ' '
  0x0021: b"\x00\x1D\x00\x00", # '!'
  0x0022: b"\x18\x00\x18\x00", # '"'
  0x0023: b"\x0E\x1B\x0E\x00", # '#'
  0x0024: b"\x0D\x1F\x16\x00", # '$'
  0x0025: b"\x12\x04\x09\x00", # '%'
  0x0026: b"\x0A\x15\x0B\x00", # '&'
  0x0027: b"\x00\x18\x00\x00", # '''
  0x0028: b"\x0E\x11\x00\x00", # '('
  0x0029: b"\x00\x11\x0E\x00", # ')'
  0x002A: b"\x0A\x04\x0A\x00", # '*'
  0x002B: b"\x04\x0E\x04\x00", # '+'
  0x002C: b"\x01\x02\x00\x00", # ','
  0x002D: b"\x04\x04\x04\x00", # '-'
  0x002E: b"\x00\x01\x00\x00", # '.'
  0x002F: b"\x02\x04\x08\x00", # '/'
  0x0030: b"\x1F\x11\x1F\x00", # '0'
  0x0031: b"\x00\x08\x1F\x00", # '1'
  0x0032: b"\x17\x15\x1D\x00", # '2'
  0x0033: b"\x15\x15\x1F\x00", # '3'
  0x0034: b"\x1C\x04\x1F\x00", # '4'
  0x0035: b"\x1D\x15\x17\x00", # '5'
  0x0036: b"\x1F\x15\x17\x00", # '6'
  0x0037: b"\x10\x10\x1F\x00", # '7'
  0x0038: b"\x1F\x15\x1F\x00", # '8'
  0x0039: b"\x1D\x15\x1F\x00", # '9'
  0x003A: b"\x00\x0A\x00\x00", # ':'
  0x003B: b"\x01\x0A\x00\x00", # ';'
  0x003C: b"\x04\x0A\x11\x00", # '<'
  0x003D: b"\x0A\x0A\x0A\x00", # '='
  0x003E: b"\x11\x0A\x04\x00", # '>'
  0x003F: b"\x18\x13\x1C\x00", # '?'
  0x0040: b"\x1C\x14\x1C\x00", # '@'
  0x0041: b"\x0F\x14\x1F\x00", # 'A'
  0x0042: b"\x1F\x15\x0A\x00", # 'B'
  0x0043: b"\x0E\x11\x11\x00", # 'C'
  0x0044: b"\x1F\x11\x0E\x00", # 'D'
  0x0045: b"\x1F\x15\x15\x00", # 'E'
  0x0046: b"\x1F\x14\x14\x00", # 'F'
  0x0047: b"\x0E\x11\x13\x00", # 'G'
  0x0048: b"\x1F\x04\x1F\x00", # 'H'
  0x0049: b"\x11\x1F\x11\x00", # 'I'
  0x004A: b"\x11\x1E\x10\x00", # 'J'
  0x004B: b"\x1F\x04\x1B\x00", # 'K'
  0x004C: b"\x1F\x01\x01\x00", # 'L'
  0x004D: b"\x1F\x0C\x1F\x00", # 'M'
  0x004E: b"\x1F\x10\x0F\x00", # 'N'
  0x004F: b"\x1F\x11\x1F\x00", # 'O'
  0x0050: b"\x1F\x14\x1C\x00", # 'P'
  0x0051: b"\x0E\x11\x0F\x00", # 'Q'
  0x0052: b"\x1F\x14\x0B\x00", # 'R'
  0x0053: b"\x09\x15\x12\x00", # 'S'
  0x0054: b"\x10\x1F\x10\x00", # 'T'
  0x0055: b"\x1E\x01\x1F\x00", # 'U'
  0x0056: b"\x1E\x01\x1E\x00", # 'V'
  0x0057: b"\x1F\x06\x1F\x00", # 'W'
  0x0058: b"\x1B\x04\x1B\x00", # 'X'
  0x0059: b"\x18\x07\x18\x00", # 'Y'
  0x005A: b"\x13\x15\x19\x00", # 'Z'
  0x005B: b"\x1F\x11\x00\x00", # '['
  0x005C: b"\x08\x04\x02\x00", # '\'
  0x005D: b"\x00\x11\x1F\x00", # ']'
  0x005E: b"\x08\x10\x08\x00", # '^'
  0x005F: b"\x01\x01\x01\x00", # '_'
  0x0060: b"\x00\x10\x08\x00", # '`'
  0x0061: b"\x12\x15\x0F\x00", # 'a'
  0x0062: b"\x1F\x05\x02\x00", # 'b'
  0x0063: b"\x02\x05\x05\x00", # 'c'
  0x0064: b"\x02\x05\x1F\x00", # 'd'
  0x0065: b"\x06\x0B\x05\x00", # 'e'
  0x0066: b"\x00\x0F\x14\x00", # 'f'
  0x0067: b"\x08\x15\x1E\x00", # 'g'
  0x0068: b"\x1F\x04\x03\x00", # 'h'
  0x0069: b"\x00\x16\x01\x00", # 'i'
  0x006A: b"\x02\x01\x16\x00", # 'j'
  0x006B: b"\x1F\x04\x0B\x00", # 'k'
  0x006C: b"\x00\x1E\x01\x00", # 'l'
  0x006D: b"\x07\x06\x03\x00", # 'm'
  0x006E: b"\x07\x04\x03\x00", # 'n'
  0x006F: b"\x07\x05\x07\x00", # 'o'
  0x0070: b"\x0F\x0A\x0E\x00", # 'p'
  0x0071: b"\x04\x0A\x0F\x00", # 'q'
  0x0072: b"\x03\x04\x04\x00", # 'r'
  0x0073: b"\x0D\x0B\x00\x00", # 's'
  0x0074: b"\x08\x1E\x09\x00", # 't'
  0x0075: b"\x06\x01\x07\x00", # 'u'
  0x0076: b"\x06\x01\x06\x00", # 'v'
  0x0077: b"\x07\x03\x07\x00", # 'w'
  0x0078: b"\x05\x02\x05\x00", # 'x'
  0x0079: b"\x04\x03\x04\x00", # 'y'
  0x007A: b"\x0B\x0D\x09\x00", # 'z'
  0x007B: b"\x04\x1B\x11\x00", # '{'
  0x007C: b"\x00\x1F\x00\x00", # '|'
  0x007D: b"\x11\x1B\x04\x00", # '}'
  0x007E: b"\x08\x18\x10\x00", # '~'
  0x00B0: b"\x08\x14\x08\x00", # '°'
  0x00B4: b"\x00\x08\x10\x00", # '´'
  0x00B7: b"\x00\x04\x00\x00", # '·'
}

def f(v, y):
    retv = ""
    for i in range(4):
        retv += "()" if (v[i] & (0x20 >> y)) != 0 else "  "
    return '"' + retv + '"'

if True: # demo mode
    print("font_4x6_src = {")

    for k, v in font_4x6.items():
        print("  0x%04X:   # '%s'" % (k, chr(k)))
     
        print('  (' + f(v, 0) + ',')
        for i in range(1, 5):
            print('   ' + f(v, i) + ',')
        print('   ' + f(v, 5) + ')')
        print()

    print("}")



font_4x6_old.py
#!/usr/bin/env python3
# -*- coding: UTF8 -*-

font_4x6 = {
  0x0020: b"\x00\x00\x00\x00\x00\x00", # ' '
  0x0021: b"\x00\x04\x04\x04\x00\x04", # '!'
  0x0022: b"\x00\x0A\x0A\x00\x00\x00", # '"'
  0x0023: b"\x00\x04\x0E\x0A\x0E\x04", # '#'
  0x0024: b"\x00\x06\x0C\x0E\x06\x0C", # '$'
  0x0025: b"\x00\x08\x02\x04\x08\x02", # '%'
  0x0026: b"\x00\x04\x0A\x04\x0A\x06", # '&'
  0x0027: b"\x00\x04\x04\x00\x00\x00", # '''
  0x0028: b"\x00\x04\x08\x08\x08\x04", # '('
  0x0029: b"\x00\x04\x02\x02\x02\x04", # ')'
  0x002A: b"\x00\x00\x0A\x04\x0A\x00", # '*'
  0x002B: b"\x00\x00\x04\x0E\x04\x00", # '+'
  0x002C: b"\x00\x00\x00\x00\x04\x08", # ','
  0x002D: b"\x00\x00\x00\x0E\x00\x00", # '-'
  0x002E: b"\x00\x00\x00\x00\x00\x04", # '.'
  0x002F: b"\x00\x00\x02\x04\x08\x00", # '/'
  0x0030: b"\x00\x0E\x0A\x0A\x0A\x0E", # '0'
  0x0031: b"\x00\x02\x06\x02\x02\x02", # '1'
  0x0032: b"\x00\x0E\x02\x0E\x08\x0E", # '2'
  0x0033: b"\x00\x0E\x02\x0E\x02\x0E", # '3'
  0x0034: b"\x00\x0A\x0A\x0E\x02\x02", # '4'
  0x0035: b"\x00\x0E\x08\x0E\x02\x0E", # '5'
  0x0036: b"\x00\x0E\x08\x0E\x0A\x0E", # '6'
  0x0037: b"\x00\x0E\x02\x02\x02\x02", # '7'
  0x0038: b"\x00\x0E\x0A\x0E\x0A\x0E", # '8'
  0x0039: b"\x00\x0E\x0A\x0E\x02\x0E", # '9'
  0x003A: b"\x00\x00\x04\x00\x04\x00", # ':'
  0x003B: b"\x00\x00\x04\x00\x04\x08", # ';'
  0x003C: b"\x00\x02\x04\x08\x04\x02", # '<'
  0x003D: b"\x00\x00\x0E\x00\x0E\x00", # '='
  0x003E: b"\x00\x08\x04\x02\x04\x08", # '>'
  0x003F: b"\x00\x0E\x0A\x02\x04\x04", # '?'
  0x0040: b"\x00\x0E\x0A\x0E\x00\x00", # '@'
  0x0041: b"\x00\x06\x0A\x0E\x0A\x0A", # 'A'
  0x0042: b"\x00\x0C\x0A\x0C\x0A\x0C", # 'B'
  0x0043: b"\x00\x06\x08\x08\x08\x06", # 'C'
  0x0044: b"\x00\x0C\x0A\x0A\x0A\x0C", # 'D'
  0x0045: b"\x00\x0E\x08\x0E\x08\x0E", # 'E'
  0x0046: b"\x00\x0E\x08\x0E\x08\x08", # 'F'
  0x0047: b"\x00\x06\x08\x08\x0A\x06", # 'G'
  0x0048: b"\x00\x0A\x0A\x0E\x0A\x0A", # 'H'
  0x0049: b"\x00\x0E\x04\x04\x04\x0E", # 'I'
  0x004A: b"\x00\x0E\x04\x04\x04\x08", # 'J'
  0x004B: b"\x00\x0A\x0A\x0C\x0A\x0A", # 'K'
  0x004C: b"\x00\x08\x08\x08\x08\x0E", # 'L'
  0x004D: b"\x00\x0A\x0E\x0E\x0A\x0A", # 'M'
  0x004E: b"\x00\x0C\x0A\x0A\x0A\x0A", # 'N'
  0x004F: b"\x00\x04\x0A\x0A\x0A\x04", # 'O'
  0x0050: b"\x00\x0E\x0A\x0E\x08\x08", # 'P'
  0x0051: b"\x00\x04\x0A\x0A\x0A\x06", # 'Q'
  0x0052: b"\x00\x0C\x0A\x0C\x0A\x0A", # 'R'
  0x0053: b"\x00\x06\x08\x04\x02\x0C", # 'S'
  0x0054: b"\x00\x0E\x04\x04\x04\x04", # 'T'
  0x0055: b"\x00\x0A\x0A\x0A\x0A\x06", # 'U'
  0x0056: b"\x00\x0A\x0A\x0A\x0A\x04", # 'V'
  0x0057: b"\x00\x0A\x0A\x0E\x0E\x0A", # 'W'
  0x0058: b"\x00\x0A\x0A\x04\x0A\x0A", # 'X'
  0x0059: b"\x00\x0A\x0A\x04\x04\x04", # 'Y'
  0x005A: b"\x00\x0E\x02\x04\x08\x0E", # 'Z'
  0x005B: b"\x00\x0C\x08\x08\x08\x0C", # '['
  0x005C: b"\x00\x00\x08\x04\x02\x00", # '\'
  0x005D: b"\x00\x06\x02\x02\x02\x06", # ']'
  0x005E: b"\x00\x04\x0A\x00\x00\x00", # '^'
  0x005F: b"\x00\x00\x00\x00\x00\x0E", # '_'
  0x0060: b"\x00\x04\x02\x00\x00\x00", # '`'
  0x0061: b"\x00\x0C\x02\x06\x0A\x06", # 'a'
  0x0062: b"\x00\x08\x08\x0C\x0A\x0C", # 'b'
  0x0063: b"\x00\x00\x00\x06\x08\x06", # 'c'
  0x0064: b"\x00\x02\x02\x06\x0A\x06", # 'd'
  0x0065: b"\x00\x00\x04\x0A\x0C\x06", # 'e'
  0x0066: b"\x00\x02\x04\x06\x04\x04", # 'f'
  0x0067: b"\x00\x06\x0A\x06\x02\x04", # 'g'
  0x0068: b"\x00\x08\x08\x0C\x0A\x0A", # 'h'
  0x0069: b"\x00\x04\x00\x04\x04\x02", # 'i'
  0x006A: b"\x00\x02\x00\x02\x0A\x04", # 'j'
  0x006B: b"\x00\x08\x0A\x0C\x0A\x0A", # 'k'
  0x006C: b"\x00\x04\x04\x04\x04\x02", # 'l'
  0x006D: b"\x00\x00\x00\x0C\x0E\x0A", # 'm'
  0x006E: b"\x00\x00\x00\x0C\x0A\x0A", # 'n'
  0x006F: b"\x00\x00\x00\x0E\x0A\x0E", # 'o'
  0x0070: b"\x00\x00\x0E\x0A\x0E\x08", # 'p'
  0x0071: b"\x00\x00\x06\x0A\x06\x02", # 'q'
  0x0072: b"\x00\x00\x00\x06\x08\x08", # 'r'
  0x0073: b"\x00\x00\x0C\x08\x04\x0C", # 's'
  0x0074: b"\x00\x04\x0E\x04\x04\x02", # 't'
  0x0075: b"\x00\x00\x00\x0A\x0A\x06", # 'u'
  0x0076: b"\x00\x00\x00\x0A\x0A\x04", # 'v'
  0x0077: b"\x00\x00\x00\x0A\x0E\x0E", # 'w'
  0x0078: b"\x00\x00\x00\x0A\x04\x0A", # 'x'
  0x0079: b"\x00\x00\x00\x0A\x04\x04", # 'y'
  0x007A: b"\x00\x00\x0E\x04\x08\x0E", # 'z'
  0x007B: b"\x00\x06\x04\x08\x04\x06", # '{'
  0x007C: b"\x00\x04\x04\x04\x04\x04", # '|'
  0x007D: b"\x00\x0C\x04\x02\x04\x0C", # '}'
  0x007E: b"\x00\x06\x0C\x00\x00\x00", # '~'
  0x00B0: b"\x00\x04\x0A\x04\x00\x00", # '°'
  0x00B4: b"\x00\x02\x04\x00\x00\x00", # '´'
  0x00B7: b"\x00\x00\x00\x04\x00\x00", # '·'
}

def f(b):
    retv = ""
    for i in range(4):
        retv += "()" if (b & (0x08 >> i)) != 0 else "  "
    return '"' + retv + '"'

if True: # demo mode
    print("font_4x6_src = {")

    for k, v in font_4x6.items():
        print("  0x%04X:   # '%s'" % (k, chr(k)))
     
        print('  (' + f(v[0]) + ',')
        for i in range(1, 5):
            print('   ' + f(v[i]) + ',')
        print('   ' + f(v[5]) + ')')
        print()

    print("}")



font_4x6_src.py
#!/usr/bin/env python3
# -*- coding: UTF8 -*-

font_4x6_src = {
  0x0020: # ' ' "\u0020"
   ("        ",
    "        ",
    "        ",
    "        ",
    "        ",
    "        "),

  0x0021: # '!' "\u0021"
   ("        ",
    "  ()    ",
    "  ()    ",
    "  ()    ",
    "        ",
    "  ()    "),

  0x0022: # '"' "\u0022"
   ("        ",
    "()  ()  ",
    "()  ()  ",
    "        ",
    "        ",
    "        "),

  0x0023: # '#' "\u0023"
   ("        ",
    "  ()    ",
    "()()()  ",
    "()  ()  ",
    "()()()  ",
    "  ()    "),

  0x0024: # '$' "\u0024"
   ("        ",
    "  ()()  ",
    "()()    ",
    "()()()  ",
    "  ()()  ",
    "()()    "),

  0x0025: # '%' "\u0025"
   ("        ",
    "()      ",
    "    ()  ",
    "  ()    ",
    "()      ",
    "    ()  "),

  0x0026: # '&' "\u0026"
   ("        ",
    "  ()    ",
    "()  ()  ",
    "  ()    ",
    "()  ()  ",
    "  ()()  "),

  0x0027: # ''' "\u0027"
   ("        ",
    "  ()    ",
    "  ()    ",
    "        ",
    "        ",
    "        "),

  0x0028: # '(' "\u0028"
   ("        ",
    "  ()    ",
    "()      ",
    "()      ",
    "()      ",
    "  ()    "),

  0x0029: # ')' "\u0029"
   ("        ",
    "  ()    ",
    "    ()  ",
    "    ()  ",
    "    ()  ",
    "  ()    "),

  0x002A: # '*' "\u002A"
   ("        ",
    "        ",
    "()  ()  ",
    "  ()    ",
    "()  ()  ",
    "        "),

  0x002B: # '+' "\u002B"
   ("        ",
    "        ",
    "  ()    ",
    "()()()  ",
    "  ()    ",
    "        "),

  0x002C: # ',' "\u002C"
   ("        ",
    "        ",
    "        ",
    "        ",
    "  ()    ",
    "()      "),

  0x002D: # '-' "\u002D"
   ("        ",
    "        ",
    "        ",
    "()()()  ",
    "        ",
    "        "),

  0x002E: # '.' "\u002E"
   ("        ",
    "        ",
    "        ",
    "        ",
    "        ",
    "  ()    "),

  0x002F: # '/' "\u002F"
   ("        ",
    "        ",
    "    ()  ",
    "  ()    ",
    "()      ",
    "        "),

  0x0030: # '0' "\u0030"
   ("        ",
    "()()()  ",
    "()  ()  ",
    "()  ()  ",
    "()  ()  ",
    "()()()  "),

  0x0031: # '1' "\u0031"
   ("        ",
    "    ()  ",
    "  ()()  ",
    "    ()  ",
    "    ()  ",
    "    ()  "),

  0x0032: # '2' "\u0032"
   ("        ",
    "()()()  ",
    "    ()  ",
    "()()()  ",
    "()      ",
    "()()()  "),

  0x0033: # '3' "\u0033"
   ("        ",
    "()()()  ",
    "    ()  ",
    "()()()  ",
    "    ()  ",
    "()()()  "),

  0x0034: # '4' "\u0034"
   ("        ",
    "()  ()  ",
    "()  ()  ",
    "()()()  ",
    "    ()  ",
    "    ()  "),

  0x0035: # '5' "\u0035"
   ("        ",
    "()()()  ",
    "()      ",
    "()()()  ",
    "    ()  ",
    "()()()  "),

  0x0036: # '6' "\u0036"
   ("        ",
    "()()()  ",
    "()      ",
    "()()()  ",
    "()  ()  ",
    "()()()  "),

  0x0037: # '7' "\u0037"
   ("        ",
    "()()()  ",
    "    ()  ",
    "    ()  ",
    "    ()  ",
    "    ()  "),

  0x0038: # '8' "\u0038"
   ("        ",
    "()()()  ",
    "()  ()  ",
    "()()()  ",
    "()  ()  ",
    "()()()  "),

  0x0039: # '9' "\u0039"
   ("        ",
    "()()()  ",
    "()  ()  ",
    "()()()  ",
    "    ()  ",
    "()()()  "),

  0x003A: # ':' "\u003A"
   ("        ",
    "        ",
    "  ()    ",
    "        ",
    "  ()    ",
    "        "),

  0x003B: # ';' "\u003B"
   ("        ",
    "        ",
    "  ()    ",
    "        ",
    "  ()    ",
    "()      "),

  0x003C: # '<' "\u003C"
   ("        ",
    "    ()  ",
    "  ()    ",
    "()      ",
    "  ()    ",
    "    ()  "),

  0x003D: # '=' "\u003D"
   ("        ",
    "        ",
    "()()()  ",
    "        ",
    "()()()  ",
    "        "),

  0x003E: # '>' "\u003E"
   ("        ",
    "()      ",
    "  ()    ",
    "    ()  ",
    "  ()    ",
    "()      "),

  0x003F: # '?' "\u003F"
   ("        ",
    "()()()  ",
    "()  ()  ",
    "    ()  ",
    "  ()    ",
    "  ()    "),

  0x0040: # '@' "\u0040"
   ("        ",
    "()()()  ",
    "()  ()  ",
    "()()()  ",
    "        ",
    "        "),
  
  0x0041: # 'A' "\u0041"
   ("        ",
    "  ()()  ",
    "()  ()  ",
    "()()()  ",
    "()  ()  ",
    "()  ()  "),
   
  0x0042: # 'B' "\u0042"
   ("        ",
    "()()    ",
    "()  ()  ",
    "()()    ",
    "()  ()  ",
    "()()    "),
   
  0x0043: # 'C' "\u0043"
   ("        ",
    "  ()()  ",
    "()      ",
    "()      ",
    "()      ",
    "  ()()  "),
   
  0x0044: # 'D' "\u0044"
   ("        ",
    "()()    ",
    "()  ()  ",
    "()  ()  ",
    "()  ()  ",
    "()()    "),
   
  0x0045: # 'E' "\u0045"
   ("        ",
    "()()()  ",
    "()      ",
    "()()()  ",
    "()      ",
    "()()()  "),
   
  0x0046: # 'F' "\u0046"
   ("        ",
    "()()()  ",
    "()      ",
    "()()()  ",
    "()      ",
    "()      "),
   
  0x0047: # 'G' "\u0047"
   ("        ",
    "  ()()  ",
    "()      ",
    "()      ",
    "()  ()  ",
    "  ()()  "),
   
  0x0048: # 'H' "\u0048"
   ("        ",
    "()  ()  ",
    "()  ()  ",
    "()()()  ",
    "()  ()  ",
    "()  ()  "),
   
  0x0049: # 'I' "\u0049"
   ("        ",
    "()()()  ",
    "  ()    ",
    "  ()    ",
    "  ()    ",
    "()()()  "),
   
  0x004A: # 'J' "\u004A"
   ("        ",
    "()()()  ",
    "  ()    ",
    "  ()    ",
    "  ()    ",
    "()      "),
   
  0x004B: # 'K' "\u004B"
   ("        ",
    "()  ()  ",
    "()  ()  ",
    "()()    ",
    "()  ()  ",
    "()  ()  "),
   
  0x004C: # 'L' "\u004C"
   ("        ",
    "()      ",
    "()      ",
    "()      ",
    "()      ",
    "()()()  "),
   
  0x004D: # 'M' "\u004D"
   ("        ",
    "()  ()  ",
    "()()()  ",
    "()()()  ",
    "()  ()  ",
    "()  ()  "),
   
  0x004E: # 'N' "\u004E"
   ("        ",
    "()()    ",
    "()  ()  ",
    "()  ()  ",
    "()  ()  ",
    "()  ()  "),
   
  0x004F: # 'O' "\u004F"
   ("        ",
    "()()()  ",
    "()  ()  ",
    "()  ()  ",
    "()  ()  ",
    "()()()  "),
   
  0x0050: # 'P' "\u0050"
   ("        ",
    "()()()  ",
    "()  ()  ",
    "()()()  ",
    "()      ",
    "()      "),
   
  0x0051: # 'Q' "\u0051"
   ("        ",
    "  ()    ",
    "()  ()  ",
    "()  ()  ",
    "()  ()  ",
    "  ()()  "),
   
  0x0052: # 'R' "\u0052"
   ("        ",
    "()()    ",
    "()  ()  ",
    "()()    ",
    "()  ()  ",
    "()  ()  "),
   
  0x0053: # 'S' "\u0053"
   ("        ",
    "  ()()  ",
    "()      ",
    "  ()    ",
    "    ()  ",
    "()()    "),
   
  0x0054: # 'T' "\u0054"
   ("        ",
    "()()()  ",
    "  ()    ",
    "  ()    ",
    "  ()    ",
    "  ()    "),
   
  0x0055: # 'U' "\u0055"
   ("        ",
    "()  ()  ",
    "()  ()  ",
    "()  ()  ",
    "()  ()  ",
    "  ()()  "),
   
  0x0056: # 'V' "\u0056"
   ("        ",
    "()  ()  ",
    "()  ()  ",
    "()  ()  ",
    "()  ()  ",
    "  ()    "),
   
  0x0057: # 'W' "\u0057"
   ("        ",
    "()  ()  ",
    "()  ()  ",
    "()()()  ",
    "()()()  ",
    "()  ()  "),
   
  0x0058: # 'X' "\u0058"
   ("        ",
    "()  ()  ",
    "()  ()  ",
    "  ()    ",
    "()  ()  ",
    "()  ()  "),
   
  0x0059: # 'Y' "\u0059"
   ("        ",
    "()  ()  ",
    "()  ()  ",
    "  ()    ",
    "  ()    ",
    "  ()    "),
   
  0x005A: # 'Z' "\u005A"
   ("        ",
    "()()()  ",
    "    ()  ",
    "  ()    ",
    "()      ",
    "()()()  "),
   
  0x005B: # '[' "\u005B"
   ("        ",
    "()()    ",
    "()      ",
    "()      ",
    "()      ",
    "()()    "),
   
  0x005C: # '\' "\u005C"
   ("        ",
    "        ",
    "()      ",
    "  ()    ",
    "    ()  ",
    "        "),
   
  0x005D: # ']' "\u005D"
   ("        ",
    "  ()()  ",
    "    ()  ",
    "    ()  ",
    "    ()  ",
    "  ()()  "),
   
  0x005E: # '^' "\u005E"
   ("        ",
    "  ()    ",
    "()  ()  ",
    "        ",
    "        ",
    "        "),
   
  0x005F: # '_' "\u005F"
   ("        ",
    "        ",
    "        ",
    "        ",
    "        ",
    "()()()  "),
   
  0x0060: # '`' "\u0060"
   ("        ",
    "  ()    ",
    "    ()  ",
    "        ",
    "        ",
    "        "),

  0x0061: # 'a' "\u0061"
   ("        ",
    "()()    ",
    "    ()  ",
    "  ()()  ",
    "()  ()  ",
    "  ()()  "),
   
  0x0062: # 'b' "\u0062"
   ("        ",
    "()      ",
    "()      ",
    "()()    ",
    "()  ()  ",
    "()()    "),
   
  0x0063: # 'c' "\u0063"
   ("        ",
    "        ",
    "        ",
    "  ()()  ",
    "()      ",
    "  ()()  "),
   
  0x0064: # 'd' "\u0064"
   ("        ",
    "    ()  ",
    "    ()  ",
    "  ()()  ",
    "()  ()  ",
    "  ()()  "),
   
  0x0065: # 'e' "\u0065"
   ("        ",
    "        ",
    "  ()    ",
    "()  ()  ",
    "()()    ",
    "  ()()  "),
   
  0x0066: # 'f' "\u0066"
   ("        ",
    "    ()  ",
    "  ()    ",
    "  ()()  ",
    "  ()    ",
    "  ()    "),
   
  0x0067: # 'g' "\u0067"
   ("        ",
    "  ()()  ",
    "()  ()  ",
    "  ()()  ",
    "    ()  ",
    "  ()    "),
   
  0x0068: # 'h' "\u0068"
   ("        ",
    "()      ",
    "()      ",
    "()()    ",
    "()  ()  ",
    "()  ()  "),
   
  0x0069: # 'i' "\u0069"
   ("        ",
    "  ()    ",
    "        ",
    "  ()    ",
    "  ()    ",
    "    ()  "),
   
  0x006A: # 'j' "\u006A"
   ("        ",
    "    ()  ",
    "        ",
    "    ()  ",
    "()  ()  ",
    "  ()    "),
   
  0x006B: # 'k' "\u006B"
   ("        ",
    "()      ",
    "()      ",
    "()  ()  ",
    "()()    ",
    "()  ()  "),
   
  0x006C: # 'l' "\u006C"
   ("        ",
    "  ()    ",
    "  ()    ",
    "  ()    ",
    "  ()    ",
    "    ()  "),
   
  0x006D: # 'm' "\u006D"
   ("        ",
    "        ",
    "        ",
    "()()    ",
    "()()()  ",
    "()  ()  "),
   
  0x006E: # 'n' "\u006E"
   ("        ",
    "        ",
    "        ",
    "()()    ",
    "()  ()  ",
    "()  ()  "),
   
  0x006F: # 'o' "\u006F"
   ("        ",
    "        ",
    "        ",
    "  ()    ",
    "()  ()  ",
    "  ()    "),
   
  0x0070: # 'p' "\u0070"
   ("        ",
    "        ",
    "()()    ",
    "()  ()  ",
    "()()    ",
    "()      "),
   
  0x0071: # 'q' "\u0071"
   ("        ",
    "        ",
    "  ()()  ",
    "()  ()  ",
    "  ()()  ",
    "    ()  "),
   
  0x0072: # 'r' "\u0072"
   ("        ",
    "        ",
    "        ",
    "  ()()  ",
    "()      ",
    "()      "),
   
  0x0073: # 's' "\u0073"
   ("        ",
    "        ",
    "()()    ",
    "()      ",
    "  ()    ",
    "()()    "),
   
  0x0074: # 't' "\u0074"
   ("        ",
    "  ()    ",
    "()()()  ",
    "  ()    ",
    "  ()    ",
    "    ()  "),
   
  0x0075: # 'u' "\u0075"
   ("        ",
    "        ",
    "        ",
    "()  ()  ",
    "()  ()  ",
    "  ()()  "),
   
  0x0076: # 'v' "\u0076"
   ("        ",
    "        ",
    "        ",
    "()  ()  ",
    "()  ()  ",
    "  ()    "),
   
  0x0077: # 'w' "\u0077"
   ("        ",
    "        ",
    "        ",
    "()  ()  ",
    "()()()  ",
    "()()()  "),
   
  0x0078: # 'x' "\u0078"
   ("        ",
    "        ",
    "        ",
    "()  ()  ",
    "  ()    ",
    "()  ()  "),
   
  0x0079: # 'y' "\u0079"
   ("        ",
    "        ",
    "        ",
    "()  ()  ",
    "  ()    ",
    "  ()    "),
   
  0x007A: # 'z' "\u007A"
   ("        ",
    "        ",
    "()()()  ",
    "  ()    ",
    "()      ",
    "()()()  "),
   
  0x007B: # '{' "\u007B"
   ("        ",
    "  ()()  ",
    "  ()    ",
    "()      ",
    "  ()    ",
    "  ()()  "),
   
  0x007C: # '|' "\u007C"
   ("        ",
    "  ()    ",
    "  ()    ",
    "  ()    ",
    "  ()    ",
    "  ()    "),
   
  0x007D: # '}' "\u007D"
   ("        ",
    "()()    ",
    "  ()    ",
    "    ()  ",
    "  ()    ",
    "()()    "),
   
  0x007E: # '~' "\u007E"
   ("        ",
    "  ()()  ",
    "()()    ",
    "        ",
    "        ",
    "        "),

  0x00B0: # '°' "\u00B0" (KOI8-R: 0x9C)
   ("        ",
    "  ()    ",
    "()  ()  ",
    "  ()    ",
    "        ",
    "        "),

  0x00B4: # '´' "\u00B4"
   ("        ",
    "    ()  ",
    "  ()    ",
    "        ",
    "        ",
    "        "),

  0x00B7: # '·' "\u00B7" (KOI8-R: 0x9E)
   ("        ",
    "        ",
    "        ",
    "  ()    ",
    "        ",
    "        "),
}

def f1(s):
    retv = ""
    for i in range(4):
        retv += "()" if s[i * 2] != " " else "  "
    return '"' + retv + '"'

def f2(s):
    retv = 0
    for i in range(4):
        if s[i * 2] != " ":
            retv |= 0x8 >> i
    return "%02X" % retv

def f3(s):
    retv = ""
    for i in range(3):
        retv += "1, " if s[i * 2] != " " else "0, "
    retv += "1" if s[3 * 2] != " " else "0"
    return '(' + retv + ')'

def f4(v, x):
    retv = 0
    for i in range(6):
        if v[i][x * 2] != " ":
            retv |= 0x20 >> i
    return "%02X" % retv

v = 6 # 1, 2, 3, 4, 5, 6 or 7

if v == 1: # demo mode
    print("font_4x6_src = {")

    for k, v in font_4x6_src.items():
        print("  0x%04X:   # '%s'" % (k, chr(k)))
     
        print('   (' + f1(v[0]) + ',')
        for i in range(1, 5):
            print('    ' + f1(v[i]) + ',')
        print('    ' + f1(v[5]) + ')')
        print()

    print("}")

elif v == 2: # compact mode (bytearrays)
    print("font_4x6 = {")

    for k, v in font_4x6_src.items():
        str = '  0x%04X: b"' % k
        for i in range(6):
            str += '\\x' + f2(v[i]) 
        str += '", # \'' + chr(k) + "'"
        print(str)
    
    print("}")

elif v == 3: # matrix mode
    print("font_4x6 = {")
    
    for k, v in font_4x6_src.items():
        print("  0x%04X: # '%s'" % (k, chr(k)))
     
        print('   (' + f3(v[0]) + ',')
        for i in range(1, 5):
            print('    ' + f3(v[i]) + ',')
        print('    ' + f3(v[5]) + ')')
        print()

    print("}")

elif v == 4: # matrix SORT mode 
    print("font_4x6_short = (")
    
    for k in range(ord(b" "), ord(b'~') + 1):
        v = font_4x6_src[k]
        print("  # '%s' = '\\x%02X'" % (chr(k), k))
        print('   (' + f3(v[0]) + ',')
        for i in range(1, 5):
            print('    ' + f3(v[i]) + ',')
        print('    ' + f3(v[5]) + ')')
        print()

    print(")")

elif v == 5: # super COMPACT mode (bytearrays)
    print("font_4x6_compact = (")

    for k in range(ord(b" "), ord(b'~') + 1):
        v = font_4x6_src[k]

        str = '  b"'
        for i in range(6):
            str += '\\x' + f2(v[i]) 
        str += '", # \'' + chr(k) + ("' = '\\x%02X'" % k)
        print(str)
    
    print(")")

elif v == 6: # mega mode (6*4)
    print("font_4x6 = {")
    
    for k, v in font_4x6_src.items():
        str = '  0x%04X: b"' % k
        for i in range(4):
            str += '\\x' + f4(v, i) 
        str += '", # \'' + chr(k) + "'"
        print(str)

    print("}")

elif v == 7: # mega mode compact (6*4)
    print("font_4x6_compact = (")
    for k in range(ord(b" "), ord(b'~') + 1):
        v = font_4x6_src[k]

        str = '  b"'
        for i in range(4):
            str += '\\x' + f4(v, i) 
        str += '", # \'' + chr(k) + "'"
        print(str)

    print(")")

